torch/_functorch/partitioners.py

```diff
@@ -1295,7 +1478,148 @@ def choose_saved_values_set(
+        return sum([_size_of(i) for i in saved_values]) / 1e9
+
+    min_act_size = estimate_activations_size(node_info.inputs)
+    max_act_size = estimate_activations_size(runtime_optimized_saved_values)
+    # The optimized choice is smaller than the inputs anyways
+    if max_act_size <= min_act_size:
+        return runtime_optimized_saved_values
+
+    def get_normalized_size(sz):
+        return (sz / 1e9) / (max_act_size - min_act_size)
+
+    def get_mem_ratio(activations: List[fx.Node]):
+        return (estimate_activations_size(activations) - min_act_size) / (
+            max_act_size - min_act_size
+        )
+
+    more_aggressive_options = replace(
+        min_cut_options,
+        ban_if_used_far_apart=False,
+        ban_if_long_fusible_chains=False,
+        ban_if_materialized_backward=False,
+    )
+    more_aggressive_saved_values, _ = get_saved_values(
+        joint_graph, node_info, more_aggressive_options
+    )
+    if get_mem_ratio(more_aggressive_saved_values) < memory_budget:
+        return more_aggressive_saved_values
+
+    aggressive_options = replace(
+        more_aggressive_options,
+        ban_if_not_in_allowlist=False,
+    )
+    aggressive_recomputation_saved_values, banned_nodes = get_saved_values(
+        joint_graph, node_info, aggressive_options
+    )
+
+    if get_mem_ratio(aggressive_recomputation_saved_values) < memory_budget:
+        return aggressive_recomputation_saved_values
+
+    from torch._inductor.fx_utils import get_node_storage
+
+    input_storages = {get_node_storage(node) for node in node_info.inputs}
+
+    def get_recomputable_banned_nodes(banned_nodes, saved_values):
+        return [
+            i
+            for i in banned_nodes
+            if (i.dist_from_bw < int(1e9) and get_node_storage(i) not in input_storages)
+        ]
+
+    recomputable_banned_nodes = get_recomputable_banned_nodes(
+        banned_nodes, aggressive_recomputation_saved_values
+    )
+
+    # default: runtime_optimized_saved_values
+    # more aggressive: more_aggressive_saved_values
+    # full aggressive: aggressive_recomputation_saved_values
+
+    all_recomputable_banned_nodes = sorted(
+        recomputable_banned_nodes, key=_size_of, reverse=True
+    )
+    if len(all_recomputable_banned_nodes) == 0:
+        return node_info.inputs
+    memories = [get_normalized_size(_size_of(i)) for i in all_recomputable_banned_nodes]
+    runtimes = [estimate_runtime(node) for node in all_recomputable_banned_nodes]
+    cur_memory_budget = memory_budget
+    from torch.utils._mode_utils import no_dispatch
+
+    def get_saved_values_knapsack(memory_budget):
+        with no_dispatch():
+            (
+                expected_runtime,
+                saved_node_idxs,
+                recomputable_node_idxs,
+            ) = _optimize_runtime_with_given_memory(
+                memories, runtimes, max(memory_budget, 0)
+            )
+        dont_ban = set()
```

Comment:
Will there be any conflict of 
- don't ban set here
- and the decision to ban a recomputation in the min-cut solver

Looks like the decision here will win. Will that be an issue?

Reply:
The logic here is that we start with a set of `banned_nodes` produced by the min-cut partitioner (i.e. compute-intensive ops). Then, the knapsack partitioner "unbans" a set of nodes from the `banned_nodes`.

Reply:
I see.

My understanding is we have the saved set of nodes (call it S), and ban-ed nodes (call it set B). B is a subset of S.

And then we pick a subset to unban from B with the knapsack solver, call it UB.

Then the set UB will be recomputed, but the remaining nodes in B (i.e. B - UB) will be saved. 

The problem I see here is, we pick `B - UB` using the memory_budget. But `B - UB` is only a subset of the nodes being saved. There are more nodes to be saved (i.e. S - B). Then we will exceed the memory budget.

It may be fine if memory_budget is only considered as a hint rather than some strict constraint.

Reply:
I agree that `memory_budget` should not be considered an *absolute* constraint, although we do aim to stick very closely to it.

Some notes:
1. It is *not* true that `B` is a subset of `S`. There may be nodes that are banned from recomputation that are nevertheless not saved. For example, imagine a matmul that's present in the forwards pass where you don't actually need its output for the backwards pass.

2. I think it's true that there are a variety of cases where we might not end up with totally the right thing. For example, if we have two compute ops that depend on each other but only the second op is required for backwards. But I think the current approach works well enough for the cases we care about.